Unit tests for engine.beginTimer

JoergAtGithub opened the pull request:
We've several reports about suspicious about the scripting API function `engine.beginTimer` in 2.4. To cover this functionality, I added testcase for this function.

JoergAtGithub committed:
Added engine.beginTimer tests

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.4' into beginTimerTest

Swiftb0y commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +620,277 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+
+    // Single shot timer with minimum allowed interval of 20ms
+    EXPECT_TRUE(evaluateAndAssert(
+            R"(var globVar = 7;
```
lets use `let` everywhere for predictable scoping rules.

JoergAtGithub replied:
I think `var` is required here - if I change this to `let`, the tests fail

Swiftb0y replied:
thats a little weird, what error?

JoergAtGithub replied:
If I change only line 691 from `var `to `let`, I get:
```
 ControllerScriptEngineLegacyTest.beginTimer_singleShotTimerBindFunction
   Source: controllerscriptenginelegacy_test.cpp line 676
   Duration: 72 ms

  Message: 

warning [Main] "Script tried setting ([Test], co) to NotANumber (NaN)"
#1 - Expected equality of these values:
  7.0
    Which is: 7
  co->get()
    Which is: 0
warning [Main] "Script tried setting ([Test], co) to NotANumber (NaN)"
#2 - Expected equality of these values:
  8.0
    Which is: 8
  co->get()
    Which is: 0

  Stack Trace: 
#1 - controllerscriptenginelegacy_test.cpp:707 line 707
#2 - controllerscriptenginelegacy_test.cpp:711 line 711

```

Swiftb0y commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +620,277 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  engine.setValue('[Test]', 'coTimerId', this.timerId);
+               }
+            }
+            var MyMapping = new MyClass;
```
```suggestion
            var MyMapping = new MyClass();
```

Swiftb0y commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +620,277 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+}
+
+TEST_F(ControllerScriptEngineLegacyTest, beginTimer_singleShotTimerArrowFunctionInClass) {
+    auto co = std::make_unique<ControlPotmeter>(ConfigKey("[Test]", "co"),
+            -10.0,
+            10.0);
+    co->setParameter(0.0);
+    auto coTimerId = std::make_unique<ControlPotmeter>(ConfigKey("[Test]", "coTimerId"),
+            -10.0,
+            50.0);
+    coTimerId->setParameter(0.0);
+    EXPECT_TRUE(evaluateAndAssert(
+            "engine.setValue('[Test]', 'co', 0.0);"));
+    EXPECT_DOUBLE_EQ(0.0, co->get());
```
can you deduplicate these sections into a test fixture?

Swiftb0y commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +620,277 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  engine.setValue('[Test]', 'coTimerId', this.timerId + 20);
+               }
+               runTimer() {
+                  this.timerId = engine.beginTimer(20, ()=>this.stopTimer(), false);                  
```
```suggestion
                  this.timerId = engine.beginTimer(20, this.stopTimer, false);                  
```
This is more interesting IMO.

Swiftb0y reviewed (commented):
have you also considered testing that `engine.beginTimer(function(){...})` work correctly (eg ensuring that `this` inside that function actually does correspond to the anonymous function object)? What about the legacy `engine.beginTimer("string code")` case?

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.4' into beginTimerTest

JoergAtGithub committed:
Changed var to let where possible

daschuer commented:
@JoergAtGithub can we remove this form the 2.4 Milestone or will you find time to address the review comments soon?

JoergAtGithub committed:
MyClass -> MyClass()

daschuer commented:
@Swiftb0y Is this now good to merge?

JoergAtGithub commented:
I've not addressed everything yet

JoergAtGithub committed:
Merge remote-tracking branch 'upstream/2.5' into beginTimerTest

JoergAtGithub committed:
Deduplicated CO initialization into inherited test class, which also improves test tree structure

JoergAtGithub committed:
Added timer test for "this.stopTimer" case

JoergAtGithub commented:
All remarks addressed now!

daschuer commented:
@Swiftb0y Can we merge this?

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,245 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+    // Single shot timer with minimum allowed interval of 20ms
+    EXPECT_TRUE(evaluateAndAssert(
+            R"(
+            class MyClass {
+               constructor() {
+                  this.timerId = undefined;
+                  this.globVar = 7;
+               }
+               stopTimer() {
+                  engine.stopTimer(this.timerId);
+                  this.timerId = 0;
+                  engine.setValue('[Test]', 'coTimerId', this.timerId + 20);
+               }
+               runTimer() {
+                  this.timerId = engine.beginTimer(20, this.stopTimer, false);              
+                  engine.setValue('[Test]', 'co', this.globVar);
+                  engine.setValue('[Test]', 'coTimerId', this.timerId);
+               }
+            }
```
This... does not look right. Specifically, the `this` inside the `stopTimer` function will not be an instance of `MyClass`, but will be the [`globalObject()`](https://doc.qt.io/qt-6/qjsvalue.html#call) instead.

This is because `this.stopTimer` is passed as an unbound function on line 859, and `beginTimer` will ultimately use [`QJSValue::call`](https://doc.qt.io/qt-6/qjsvalue.html#call) for invoking the callback.

Swiftb0y replied:
makes sense. it would indeed need to be
```suggestion
            class MyClass {
               constructor() {
                  this.timerId = undefined;
                  this.globVar = 7;
               }
               stopTimer() {
                  engine.stopTimer(this.timerId);
                  this.timerId = 0;
                  engine.setValue('[Test]', 'coTimerId', this.timerId + 20);
               }
               runTimer() {
                  this.timerId = engine.beginTimer(20, this.stopTimer.bind(this), false);              
                  engine.setValue('[Test]', 'co', this.globVar);
                  engine.setValue('[Test]', 'coTimerId', this.timerId);
               }
            }
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,245 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  this.globVar = 7;
+               }
+               runTimer() {
+                  this.timerId = engine.beginTimer(20, ()=> {
```
Small formatting nitpick.
```suggestion
                  this.timerId = engine.beginTimer(20, () => {
```

cr7pt0gr4ph7 replied:
Also, it would make sense to explicitly check whether `this` has actually the value we expect here (using a hypothetical `assertThat` function):

```suggestion
                  const savedThis = this;
                  this.timerId = engine.beginTimer(20, () => {
                     assertThat(savedThis === this);
                     assertThat(this instanceof MyClass);
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,245 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  this.timerId = undefined;
+                  this.globVar = 7;
+               }
+               stopTimer() {
```
It would make sense to explicitly check whether `this` has actually the value we expect here (using a hypothetical `assertThat` function):

```suggestion
               stopTimer() {
                   assertThat(this instanceof MyClass);
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,245 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  this.timerId = undefined;
+                  this.globVar = 7;
+               }
+               stopTimer() {
```
It would make sense to explicitly check whether `this` has actually the value we expect here (using a hypothetical `assertThat` function):

```suggestion
               stopTimer() {
                   assertThat(this instanceof MyClass);
```

JoergAtGithub committed:
Use `this.stopTimer.bind(this)` for correct context preservation when called as a callback.
Introduce checks to ensure `this` is an instance of `MyClass` in stopTimer methods.

daschuer commented:
@cr7pt0gr4ph7 Are you satisfied with the changes and can approve the review? Than we can merge it as mart of 2.5.1 release.

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,250 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  this.globVar = 7;
+               }
+               stopTimer() {    
+                  if (!(this instanceof MyClass)) {throw new Error("this should be an instance of MyClass");}
```
Small formatting nitpick.
```suggestion
                  if (!(this instanceof MyClass)) { throw new Error("this should be an instance of MyClass"); }
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,250 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  engine.setValue('[Test]', 'coTimerId', this.timerId + 20);
+               }
+               runTimer() {
+                  this.timerId = engine.beginTimer(20, ()=>this.stopTimer(), false);                  
```
Small formatting nitpick.
```suggestion
                  this.timerId = engine.beginTimer(20, () => this.stopTimer(), false);                  
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,250 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+                  this.globVar = 7;
+               }
+               stopTimer() {
+                  if (!(this instanceof MyClass)) {throw new Error("this should be an instance of MyClass");}
```
Small formatting nitpick.
```suggestion
                  if (!(this instanceof MyClass)) { throw new Error("this should be an instance of MyClass"); }
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,250 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+               runTimer() {                  
+                  const savedThis = this;
+                  this.timerId = engine.beginTimer(20, () => {
+                     if (savedThis !== this) {throw new Error("savedThis should be equal to this");}
+                     if (!(this instanceof MyClass)) {throw new Error("this should be an instance of MyClass");}
```
Small formatting nitpicks.
```suggestion
                     if (savedThis !== this) { throw new Error("savedThis should be equal to this"); }
                     if (!(this instanceof MyClass)) { throw new Error("this should be an instance of MyClass"); }
```

cr7pt0gr4ph7 commented on src/test/controllerscriptenginelegacy_test.cpp:
```diff
@@ -620,3 +636,250 @@ TEST_F(ControllerScriptEngineLegacyTest, connectionExecutesWithCorrectThisObject
+    // Single shot timer with minimum allowed interval of 20ms
+    EXPECT_TRUE(evaluateAndAssert(
+            R"(var globVar = 7;
+            timerId = engine.beginTimer(20, ()=> {
```
Small formatting nitpick.
```suggestion
            timerId = engine.beginTimer(20, () => {
```

cr7pt0gr4ph7 reviewed (changes requested):
LGTM, except for a few whitespace nitpicks. Looks good otherwise. Thanks!

@daschuer Sorry for taking a long time to get back to this, I've been a bit out of the loop and did not notice your comment sooner.

JoergAtGithub commented:
@cr7pt0gr4ph7 Adding spaces as suggested are formatting changes. clang-format will revert them when I commit them.

cr7pt0gr4ph7 commented:
> @cr7pt0gr4ph7 Adding spaces as suggested are formatting changes. clang-format will revert them when I commit them.

Huh, that's strange. `clang-format` shouldn't touch the raw string literals, and I can't reproduce the problem when running `clang-format` locally via the `pre-commit` hooks.

daschuer commented:
The CI is also running pre-commit with clang-format and it is not complaining. 
https://github.com/mixxxdj/mixxx/actions/runs/13378464720/job/37362545262?pr=12437
Unfortunately the log is expired but it is clang-format 19.1.3.
Which one is yours?

JoergAtGithub committed:
Update src/test/controllerscriptenginelegacy_test.cpp

Co-authored-by: Lukas Waslowski <cr7pt0gr4ph7@gmail.com>

JoergAtGithub committed:
Update src/test/controllerscriptenginelegacy_test.cpp

Co-authored-by: Lukas Waslowski <cr7pt0gr4ph7@gmail.com>

JoergAtGithub committed:
Update src/test/controllerscriptenginelegacy_test.cpp

Co-authored-by: Lukas Waslowski <cr7pt0gr4ph7@gmail.com>

JoergAtGithub committed:
Update src/test/controllerscriptenginelegacy_test.cpp

Co-authored-by: Lukas Waslowski <cr7pt0gr4ph7@gmail.com>

JoergAtGithub commented:
I don't think so

cr7pt0gr4ph7 reviewed (approved):
Looks fine to me :+1: Thanks!

JoergAtGithub commented:
From my side it's fine too!

daschuer reviewed (approved):
Thank you.
